src/components/ReanimatedSwipeable.tsx

```diff
@@ -391,17 +396,34 @@ const Swipeable = forwardRef<SwipeableMethods, SwipeableProps>(
           ...animationOptionsProp,
         };
 
+        const isClosing = toValue === 0;
+        const moveToRight = isClosing ? rowState.value < 0 : toValue > 0;
+
+        const usedWidth = isClosing
+          ? moveToRight
+            ? rightWidth.value
+            : leftWidth.value
+          : moveToRight
+          ? leftWidth.value
+          : rightWidth.value;
+
         const progressSpringConfig = {
           ...translationSpringConfig,
-          velocity: 0,
+          restDisplacementThreshold: 0.01,
+          restSpeedThreshold: 0.01,
+          velocity:
+            velocityX &&
+            interpolate(velocityX, [-usedWidth, usedWidth], [-1, 1]),
         };
 
+        const frozenRowState = rowState.value;
+
         appliedTranslation.value = withSpring(
           toValue,
           translationSpringConfig,
           (isFinished) => {
             if (isFinished) {
-              runOnJS(dispatchEndEvents)(fromValue, toValue);
+              runOnJS(dispatchEndEvents)(frozenRowState, toValue);
```

Comment:
Are all events executed on the JS thread?

Reply:
I think we can update the calling logic to check whether the callback is a worklet and run it on the UI thread if it is (that's for another PR though).

Reply:
Great idea, will do that.